snapcraft/_store.py

```diff
@@ -793,19 +793,74 @@ def validate(snap_name, validations, revoke=False, key=None):
         if revoke:
             assertion['revoked'] = "true"
 
-        assertion = _sign_validation(validation, assertion, key)
+        assertion = _sign_assertion(validation, assertion, key, 'validations')
 
         # Save assertion to a properly named file
         fname = '{}-{}-r{}.assertion'.format(snap_name, gated_name, rev)
         with open(fname, 'wb') as f:
             f.write(assertion)
 
-        store.push_validation(snap_id, assertion)
+        store.push_assertion(snap_id, assertion, endpoint='validations')
 
 
 validation_re = re.compile('^[^=]+=[0-9]+$')
 
 
+def collaborate(snap_name, key):
+    store = storeapi.StoreClient()
+
+    with _requires_login():
+        account_info = store.get_account_information()
+    publisher_id = account_info['account_id']
+
+    release = storeapi.constants.DEFAULT_SERIES
+    try:
+        snap_id = account_info['snaps'][release][snap_name]['snap-id']
+    except KeyError:
+        raise storeapi.errors.SnapNotFoundError(snap_name)
+    assertion = _get_developers(snap_id, publisher_id)
+
+    # XXX: Do the amendments via UI here.
+    #
+    # The data will look like:
+    # {'snap_developer': {
+    #      'type: 'snap-developer',
+    #      'authority-id': <account_id of the publisher>,
+    #      'publisher-id': <account_id of the publisher>,
+    #      'snap-id': 'snap_id',
+    #      'developers': [{
+    #          'developer-id': 'account_id of dev-1',
+    #          'since': '2017-02-10T08:35:00.390258Z'
+    #         },{
+    #          'developer-id': 'account_id of dev-2',
+    #          'since': '2017-02-10T08:35:00.390258Z',
+    #          'until': '2018-02-10T08:35:00.390258Z'
+    #         }],
+    #      }
+    # }
+
+    # The revision should be incremented, to avoid `invalid-revision` errors.
+    assertion['revision'] = str(int(assertion.get('revision', '0'))+1)
+
+    assertion = _sign_assertion(snap_id, assertion, key, 'developers')
+
+    store.push_assertion(snap_id, assertion, 'developers')
+
+
+def _get_developers(snap_id, publisher_id):
+    store = storeapi.StoreClient()
+    try:
+        return store.get_assertion(snap_id, 'developers')['snap_developer']
```

Comment:
When the assertion comes from the store, we don't know if the `authority-id` is "canonical" or the publisher's ID so it must be set to the publisher's ID. Otherwise the signing key does not belong to the signing authority and the assertion will be rejected when it's saved.

Personally, I would do this in the same place as `revision` is incremented but I'll leave that to you.